docs/language-specification/instructions/gates.md

```diff
@@ -0,0 +1,80 @@
+`<gate-name:ID>(<parameter(s):FLOAT|INT,>)? <qubit-argument(s):QUBIT,>`
```

Comment:
This is not easy to express as a regular expression. It would be something like:

`\(<parameter:FLOAT|INT>(,<parameter:FLOAT|INT>)*\)`

To keep the reading more user friendly I would simply write:

`gate-name:ID(<parameters:[FLOAT|INT]*>) <qubit-arguments:[QUBIT]+>`

`[FLOAT|INT]*` could be read as a list of 0 or more `FLOAT` and `INT`s.
`[QUBIT]+` could be read as a list of 1 or more `QUBIT`s.

Reply:
In the case of 0 parameters, is it then clear that the parentheses shouldn't be used?

Reply:
No, then you could add `?`. But I'm not convinced at all about using regular expressions for this type of documentation. They look to me quite low level stuff. Maybe just `gate-name[(parameters)] qubit-arguments`? Saying that `[]` means that `(parameters)` are optional, and that they are a list of one or more elements of type INT or FLOAT, and the same for `qubit-arguments`?

Reply:
I would like to be as explicit as possible. According to Richard, people often came up with very strange ways of writing and wouldn't understand where they had gone wrong. I don't wont to do a compromise between explicit and simple, I would just do both then. I'll have a look.

Reply:
What I meant is that you would use regular expressions as an analyst to pass on to a developer. But I'm not sure you need that level of detail in a language specification document. In the end, if someone would want to get to that level of detail, they could always look at the grammar file itself.

Also, the examples you are using are very good, and much more informative.

Reply:
I will use the simple form and put the regex as a collapsed Note.